File: src/Clustering/View/GMUDefaultClusterRenderer.m
/* Copyright (c) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#if !defined(__has_feature) || !__has_feature(objc_arc)
#error "This file requires ARC support."
#endif

#import <GoogleMaps/GoogleMaps.h>

#import "GMUDefaultClusterRenderer.h"
#import "GMUClusterIconGenerator.h"
#import "GMUWrappingDictionaryKey.h"

// Clusters smaller than this threshold will be expanded.
static const NSUInteger kGMUMinClusterSize = 4;

// At zooms above this level, clusters will be expanded.
// This is to prevent cases where items are so close to each other than they are always grouped.
static const float kGMUMaxClusterZoom = 20;

// Animation duration for marker splitting/merging effects.
static const double kGMUAnimationDuration = 0.5;  // seconds.

@implementation GMUDefaultClusterRenderer {
  // Map view to render clusters on.
  __weak GMSMapView *_mapView;

  // Collection of markers added to the map.
  NSMutableArray<GMSMarker *> *_mutableMarkers;

  // Icon generator used to create cluster icon.
  id<GMUClusterIconGenerator> _clusterIconGenerator;

  // Current clusters being rendered.
  NSArray<id<GMUCluster>> *_clusters;

  // Tracks clusters that have been rendered to the map.
  NSMutableSet *_renderedClusters;

  // Tracks cluster items that have been rendered to the map.
  NSMutableSet *_renderedClusterItems;

  // Stores previous zoom level to determine zooming direction (in/out).
  float _previousZoom;

  // Lookup map from cluster item to an old cluster.
  NSMutableDictionary<GMUWrappingDictionaryKey *, id<GMUCluster>> *_itemToOldClusterMap;

  // Lookup map from cluster item to a new cluster.
  NSMutableDictionary<GMUWrappingDictionaryKey *, id<GMUCluster>> *_itemToNewClusterMap;
}

- (instancetype)initWithMapView:(GMSMapView *)mapView
           clusterIconGenerator:(id<GMUClusterIconGenerator>)iconGenerator {
  if ((self = [super init])) {
    _mapView = mapView;
    _mutableMarkers = [[NSMutableArray<GMSMarker *> alloc] init];
    _clusterIconGenerator = iconGenerator;
    _renderedClusters = [[NSMutableSet alloc] init];
    _renderedClusterItems = [[NSMutableSet alloc] init];
    _animatesClusters = YES;
    _minimumClusterSize = kGMUMinClusterSize;
    _maximumClusterZoom = kGMUMaxClusterZoom;
    _animationDuration = kGMUAnimationDuration;

    _zIndex = 1;
  }
  return self;
}

- (void)dealloc {
  [self clear];
}

- (BOOL)shouldRenderAsCluster:(id<GMUCluster>)cluster atZoom:(float)zoom {
  return cluster.count >= _minimumClusterSize && zoom <= _maximumClusterZoom;
}

#pragma mark GMUClusterRenderer

- (void)renderClusters:(NSArray<id<GMUCluster>> *)clusters {
  [_renderedClusters removeAllObjects];
  [_renderedClusterItems removeAllObjects];

  if (_animatesClusters) {
    [self renderAnimatedClusters:clusters];
  } else {
    // No animation, just remove existing markers and add new ones.
    _clusters = [clusters copy];
    [self clearMarkers:_mutableMarkers];
    _mutableMarkers = [[NSMutableArray<GMSMarker *> alloc] init];
    [self addOrUpdateClusters:clusters animated:NO];
  }
}

- (void)renderAnimatedClusters:(NSArray<id<GMUCluster>> *)clusters {
  float zoom = _mapView.camera.zoom;
  BOOL isZoomingIn = zoom > _previousZoom;

  [self prepareClustersForAnimation:clusters isZoomingIn:isZoomingIn];

  _previousZoom = zoom;

  _clusters = [clusters copy];

  NSArray *existingMarkers = _mutableMarkers;
  _mutableMarkers = [[NSMutableArray<GMSMarker *> alloc] init];

  [self addOrUpdateClusters:clusters animated:isZoomingIn];

  if (isZoomingIn) {
    [self clearMarkers:existingMarkers];
  } else {
    [self clearMarkersAnimated:existingMarkers];
  }
}

- (void)clearMarkersAnimated:(NSArray<GMSMarker *> *)markers {
  // Remove existing markers: animate to nearest new cluster.
  GMSCoordinateBounds *visibleBounds =
      [[GMSCoordinateBounds alloc] initWithRegion:[_mapView.projection visibleRegion]];

  for (GMSMarker *marker in markers) {
    // If the marker for the attached userData has just been added, do not perform animation.
    if ([_renderedClusterItems containsObject:marker.userData]) {
      marker.map = nil;
      continue;
    }
    // If the marker is outside the visible view port, do not perform animation.
    if (![visibleBounds containsCoordinate:marker.position]) {
      marker.map = nil;
      continue;
    }

    // Find a candidate cluster to animate to.
    id<GMUCluster> toCluster = nil;
    if ([marker.userData conformsToProtocol:@protocol(GMUCluster)]) {
      id<GMUCluster> cluster = marker.userData;
      toCluster = [self overlappingClusterForCluster:cluster itemMap:_itemToNewClusterMap];
    } else {
      GMUWrappingDictionaryKey *key =
          [[GMUWrappingDictionaryKey alloc] initWithObject:marker.userData];
      toCluster = [_itemToNewClusterMap objectForKey:key];
    }
    // If there is not near by cluster to animate to, do not perform animation.
    if (toCluster == nil) {
      marker.map = nil;
      continue;
    }

    // All is good, perform the animation.
    [CATransaction begin];
    [CATransaction setAnimationDuration:_animationDuration];
    CLLocationCoordinate2D toPosition = toCluster.position;
    marker.layer.latitude = toPosition.latitude;
    marker.layer.longitude = toPosition.longitude;
    [CATransaction commit];
  }

  // Clears existing markers after animation has presumably ended.
  dispatch_after(dispatch_time(DISPATCH_TIME_NOW, _animationDuration * NSEC_PER_SEC),
                 dispatch_get_main_queue(), ^{
                   [self clearMarkers:markers];
                 });
}

// Called when camera is changed to reevaluate if new clusters need to be displayed because
// they become visible.
- (void)update {
  [self addOrUpdateClusters:_clusters animated:NO];
}

- (NSArray<GMSMarker *> *)markers {
  return [_mutableMarkers copy];
}

#pragma mark Private

// Builds lookup map for item to old clusters, new clusters.
- (void)prepareClustersForAnimation:(NSArray<id<GMUCluster>> *)newClusters
                        isZoomingIn:(BOOL)isZoomingIn {
  float zoom = _mapView.camera.zoom;

  if (isZoomingIn) {
    _itemToOldClusterMap =
        [[NSMutableDictionary<GMUWrappingDictionaryKey *, id<GMUCluster>> alloc] init];
    for (id<GMUCluster> cluster in _clusters) {
      if (![self shouldRenderAsCluster:cluster atZoom:zoom]
          && ![self shouldRenderAsCluster:cluster atZoom:_previousZoom]) {
        continue;
      }
      for (id<GMUClusterItem> clusterItem in cluster.items) {
        GMUWrappingDictionaryKey *key =
            [[GMUWrappingDictionaryKey alloc] initWithObject:clusterItem];
        [_itemToOldClusterMap setObject:cluster forKey:key];
      }
    }
    _itemToNewClusterMap = nil;
  } else {
    _itemToOldClusterMap = nil;
    _itemToNewClusterMap =
        [[NSMutableDictionary<GMUWrappingDictionaryKey *, id<GMUCluster>> alloc] init];
    for (id<GMUCluster> cluster in newClusters) {
      if (![self shouldRenderAsCluster:cluster atZoom:zoom]) continue;
      for (id<GMUClusterItem> clusterItem in cluster.items) {
        GMUWrappingDictionaryKey *key =
            [[GMUWrappingDictionaryKey alloc] initWithObject:clusterItem];
        [_itemToNewClusterMap setObject:cluster forKey:key];
      }
    }
  }
}

// Goes through each cluster |clusters| and add a marker for it if it is:
// - inside the visible region of the camera.
// - not yet already added.
- (void)addOrUpdateClusters:(NSArray<id<GMUCluster>> *)clusters animated:(BOOL)animated {
  GMSCoordinateBounds *visibleBounds =
      [[GMSCoordinateBounds alloc] initWithRegion:[_mapView.projection visibleRegion]];

  for (id<GMUCluster> cluster in clusters) {
    if ([_renderedClusters containsObject:cluster]) continue;

    BOOL shouldShowCluster = [visibleBounds containsCoordinate:cluster.position];
    BOOL shouldRenderAsCluster = [self shouldRenderAsCluster:cluster atZoom: _mapView.camera.zoom];

    if (!shouldShowCluster) {
      for (id<GMUClusterItem> item in cluster.items) {
        if (!shouldRenderAsCluster && [visibleBounds containsCoordinate:item.position]) {
          shouldShowCluster = YES;
          break;
        }
        if (animated) {
          GMUWrappingDictionaryKey *key = [[GMUWrappingDictionaryKey alloc] initWithObject:item];
          id<GMUCluster> oldCluster = [_itemToOldClusterMap objectForKey:key];
          if (oldCluster != nil && [visibleBounds containsCoordinate:oldCluster.position]) {
            shouldShowCluster = YES;
            break;
          }
        }
      }
    }
    if (shouldShowCluster) {
      [self renderCluster:cluster animated:animated];
    }
  }
}

- (void)renderCluster:(id<GMUCluster>)cluster animated:(BOOL)animated {
  float zoom = _mapView.camera.zoom;
  if ([self shouldRenderAsCluster:cluster atZoom:zoom]) {
    CLLocationCoordinate2D fromPosition = kCLLocationCoordinate2DInvalid;
    if (animated) {
      id<GMUCluster> fromCluster =
          [self overlappingClusterForCluster:cluster itemMap:_itemToOldClusterMap];
      animated = fromCluster != nil;
      fromPosition = fromCluster.position;
    }

    UIImage *icon = [_clusterIconGenerator iconForSize:cluster.count];
    GMSMarker *marker = [self markerWithPosition:cluster.position
                                            from:fromPosition
                                        userData:cluster
                                     clusterIcon:icon
                                        animated:animated];
    [_mutableMarkers addObject:marker];
  } else {
    for (id<GMUClusterItem> item in cluster.items) {
      GMSMarker *marker;
      if ([item class] == [GMSMarker class]) {
        marker = (GMSMarker<GMUClusterItem> *)item;
        marker.map = _mapView;
      } else {
        CLLocationCoordinate2D fromPosition = kCLLocationCoordinate2DInvalid;
        BOOL shouldAnimate = animated;
        if (shouldAnimate) {
          GMUWrappingDictionaryKey *key = [[GMUWrappingDictionaryKey alloc] initWithObject:item];
          id<GMUCluster> fromCluster = [_itemToOldClusterMap objectForKey:key];
          shouldAnimate = fromCluster != nil;
          fromPosition = fromCluster.position;
        }
        marker = [self markerWithPosition:item.position
                                     from:fromPosition
                                 userData:item
                              clusterIcon:nil
                                 animated:shouldAnimate];
        if ([item respondsToSelector:@selector(title)]) {
            marker.title = item.title;
        }
        if ([item respondsToSelector:@selector(snippet)]) {
            marker.snippet = item.snippet;
        }
      }
      [_mutableMarkers addObject:marker];
      [_renderedClusterItems addObject:item];
    }
  }
  [_renderedClusters addObject:cluster];
}

- (GMSMarker *)markerForObject:(id)object {
  GMSMarker *marker;
  if ([_delegate respondsToSelector:@selector(renderer:markerForObject:)]) {
    marker = [_delegate renderer:self markerForObject:object];
  }
  return marker ?: [[GMSMarker alloc] init];
}

// Returns a marker at final position of |position| with attached |userData|.
// If animated is YES, animates from the closest point from |points|.
- (GMSMarker *)markerWithPosition:(CLLocationCoordinate2D)position
                             from:(CLLocationCoordinate2D)from
                         userData:(id)userData
                      clusterIcon:(UIImage *)clusterIcon
                         animated:(BOOL)animated {
  GMSMarker *marker = [self markerForObject:userData];
  CLLocationCoordinate2D initialPosition = animated ? from : position;
  marker.position = initialPosition;
  marker.userData = userData;
  if (clusterIcon != nil) {
    marker.icon = clusterIcon;
    marker.groundAnchor = CGPointMake(0.5, 0.5);
  }
  marker.zIndex = _zIndex;

  if ([_delegate respondsToSelector:@selector(renderer:willRenderMarker:)]) {
    [_delegate renderer:self willRenderMarker:marker];
  }
  marker.map = _mapView;

  if (animated) {
    [CATransaction begin];
    [CATransaction setAnimationDuration:_animationDuration];
    marker.layer.latitude = position.latitude;
    marker.layer.longitude = position.longitude;
    [CATransaction commit];
  }

  if ([_delegate respondsToSelector:@selector(renderer:didRenderMarker:)]) {
    [_delegate renderer:self didRenderMarker:marker];
  }
  return marker;
}

// Returns clusters which should be rendered and is inside the camera visible region.
- (NSArray<id<GMUCluster>> *)visibleClustersFromClusters:(NSArray<id<GMUCluster>> *)clusters {
  NSMutableArray *visibleClusters = [[NSMutableArray alloc] init];
  float zoom = _mapView.camera.zoom;
  GMSCoordinateBounds *visibleBounds =
      [[GMSCoordinateBounds alloc] initWithRegion:[_mapView.projection visibleRegion]];
  for (id<GMUCluster> cluster in clusters) {
    if (![visibleBounds containsCoordinate:cluster.position]) continue;
    if (![self shouldRenderAsCluster:cluster atZoom:zoom]) continue;
    [visibleClusters addObject:cluster];
  }
  return visibleClusters;
}

// Returns the first cluster in |itemMap| that shares a common item with the input |cluster|.
// Used for heuristically finding candidate cluster to animate to/from.
- (id<GMUCluster>)overlappingClusterForCluster:
    (id<GMUCluster>)cluster
        itemMap:(NSDictionary<GMUWrappingDictionaryKey *, id<GMUCluster>> *)itemMap {
  id<GMUCluster> found = nil;
  for (id<GMUClusterItem> item in cluster.items) {
    GMUWrappingDictionaryKey *key = [[GMUWrappingDictionaryKey alloc] initWithObject:item];
    id<GMUCluster> candidate = [itemMap objectForKey:key];
    if (candidate != nil) {
      found = candidate;
      break;
    }
  }
  return found;
}

// Removes all existing markers from the attached map.
- (void)clear {
  [self clearMarkers:_mutableMarkers];
  [_mutableMarkers removeAllObjects];
  [_renderedClusters removeAllObjects];
  [_renderedClusterItems removeAllObjects];
  [_itemToNewClusterMap removeAllObjects];
  [_itemToOldClusterMap removeAllObjects];
  _clusters = nil;
}

- (void)clearMarkers:(NSArray<GMSMarker *> *)markers {
  for (GMSMarker *marker in markers) {
    marker.userData = nil;
    marker.map = nil;
  }
}

@end

File: Sources/GoogleMapsUtils/Clustering/View/DefaultRenderer/GMUDefaultClusterRenderer.swift
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import GoogleMaps

/// Default cluster renderer which shows clusters as markers with specialized icons.
/// There is logic to decide whether to expand a cluster or not depending on the number of items or the zoom level.
/// There is also some performance optimization where only clusters within the visisble region are shown.
///
final class GMUDefaultClusterRenderer: GMUClusterRenderer {

    // MARK: - Properties
    // Map view to render clusters on.
    private weak var mapView: GMSMapView?
    /// Collection of markers added to the map.
    private var mutableMarkers: [GMSMarker]
    /// Icon generator used to create cluster icon.
    private var clusterIconGenerator: GMUClusterIconGenerator
    /// Current clusters being rendered.
    private var clusters: [GMUCluster]?
    // Tracks clusters that have been rendered to the map.
    private var renderedClusters: NSMutableSet
    /// Tracks cluster items that have been rendered to the map.
    private var renderedClusterItems: NSMutableSet
    /// Stores previous zoom level to determine zooming direction (in/out).
    private var previousZoom: Float = 0.0
    /// Lookup map from cluster item to an old cluster.
    private var itemToOldClusterMap: [GMUWrappingDictionaryKey: GMUCluster]?
    /// Lookup map from cluster item to a new cluster.
    private var itemToNewClusterMap: [GMUWrappingDictionaryKey: GMUCluster]?
    /// Animates the clusters to achieve splitting (when zooming in) and merging
    /// (when zooming out) effects:
    /// - splitting large clusters into smaller ones when zooming in.
    /// - merging small clusters into bigger ones when zooming out.
    ///
    /// NOTES: the position to animate to/from for each cluster is heuristically
    /// calculated by finding the first overlapping cluster. This means that:
    /// - when zooming in:
    ///    if a cluster on a higher zoom level is made from multiple clusters on
    ///    a lower zoom level the split will only animate the new cluster from
    ///    one of them.
    /// - when zooming out:
    ///    if a cluster on a higher zoom level is split into multiple parts to join
    ///    multiple clusters at a lower zoom level, the merge will only animate
    ///    the old cluster into one of them.
    ///
    /// Because of these limitations, the actual cluster sizes may not add up, for
    /// example people may see 3 clusters of size 3, 4, 5 joining to make up a cluster
    /// of only 8 for non-hierarchical clusters. And vice versa, a cluster of 8 may
    /// split into 3 clusters of size 3, 4, 5. For hierarchical clusters, the numbers
    /// should add up however.
    ///
    /// Defaults to YES.
    var animatesClusters: Bool
    /// Clusters smaller than this threshold will be expanded.
    /// Determines the minimum number of cluster items inside a cluster.
    /// Defaults to 4.
    var minimumClusterSize: Int
    /// At zooms above this level, clusters will be expanded.
    /// This is to prevent cases where items are so close to each other than they are always grouped.
    /// Sets the maximium zoom level of the map on which the clustering
    /// should be applied. At zooms above this level, clusters will be expanded.
    /// This is to prevent cases where items are so close to each other than they
    /// are always grouped.
    /// Defaults to 20.
    var maximumClusterZoom: Int
    /// Sets the animation duration for marker splitting/merging effects.
    /// Measured in seconds.
    /// Animation duration for marker splitting/merging effects.
    /// Defaults to 0.5.
    var animationDuration: Double
    /// Allows setting a zIndex value for the clusters.
    /// This becomes useful when using multiple cluster data sets on the map and require a predictable way of displaying multiple sets with a predictable layering order.
    /// If no specific zIndex is not specified during the initialization, the default zIndex is '1'.
    /// Larger zIndex values are drawn over lower ones similar to the zIndex value of GMSMarkers.
    var zIndex: Int
    /// Sets to further customize the renderer.
    weak var delegate: GMUClusterRendererDelegate?
    /// Returns currently active markers.
    var currentActivemarkers: [GMSMarker] {
        return mutableMarkers
    }

    // MARK: - Init
    /// Initializes the cluster renderer with a map view and an icon generator.
    ///
    /// - Parameters:
    ///   - mapView: The `GMSMapView` on which clusters will be rendered.
    ///   - clusterIconGenerator: The icon generator used to create icons for the clusters.
    ///
    /// - Returns: A new instance of `GMUDefaultClusterRenderer`.
    ///
    init(mapView: GMSMapView, clusterIconGenerator: GMUClusterIconGenerator) {
        // Initialize properties
        self.mapView = mapView
        self.mutableMarkers = []
        self.clusterIconGenerator = clusterIconGenerator
        self.renderedClusters = NSMutableSet()
        self.renderedClusterItems = NSMutableSet()
        self.animatesClusters = true
        self.minimumClusterSize = 4
        self.maximumClusterZoom = 20
        self.animationDuration = 0.5 // seconds.
        self.zIndex = 1
    }

    /// Deallocates the resources by clearing all markers and clusters from the map.
    ///
    deinit {
        clear()
    }

    /// Determines whether the given cluster should be rendered as a cluster based on the current zoom level
    /// and the number of items in the cluster.
    ///
    /// - Parameters:
    ///   - cluster: The `GMUCluster` to check.
    ///   - zoom: The current zoom level of the map.
    ///
    /// - Returns: `true` if the cluster should be rendered as a cluster, otherwise `false`.
    ///
    func shouldRenderAsCluster(_ cluster: GMUCluster, atZoom zoom: Float) -> Bool {
        return cluster.count >= minimumClusterSize && Int(zoom) <= maximumClusterZoom
    }

    // MARK: - `GMUClusterRenderer`
    /// Renders the provided clusters on the map. If cluster animations are enabled, it renders
    /// the clusters with animation. Otherwise, it removes existing markers and renders new ones
    /// without animation.
    ///
    /// - Parameter clusters: The clusters to be rendered on the map.
    ///
    func renderClusters(_ clusters: [GMUCluster]) {
        renderedClusters.removeAllObjects()
        renderedClusterItems.removeAllObjects()

        if animatesClusters {
            renderAnimatedClusters(clusters)
        } else {
            // No animation, just remove existing markers and add new ones.
            self.clusters = clusters
            clearMarkers(mutableMarkers)
            mutableMarkers = []
            addOrUpdateClusters(clusters, animated: false)
        }
    }

    /// Renders clusters with animation based on zoom level.
    /// If zooming in, clusters will split; if zooming out, clusters will merge.
    /// - Parameter clusters: The clusters to render.
    ///
    func renderAnimatedClusters(_ clusters: [GMUCluster]) {
        guard let mapView else { return }
        let zoom: Float = mapView.camera.zoom
        let isZoomingIn: Bool = zoom > previousZoom

        /// Prepares the clusters for animation depending on the zoom direction.
        prepareClustersForAnimation(clusters, isZoomingIn: isZoomingIn)

        previousZoom = zoom

        self.clusters = clusters
        
        var existingMarkers: [GMSMarker] = mutableMarkers
        mutableMarkers = []

        /// Adds or updates clusters with animation based on zooming direction.
        addOrUpdateClusters(clusters, animated: isZoomingIn)

        // If the marker was re-added, remove it from existingMarkers which will be cleared
        for visibleMarker in mutableMarkers {
            existingMarkers.removeAll { $0 == visibleMarker }
        }

        /// Clears markers depending on zooming direction.
        if isZoomingIn {
            clearMarkers(existingMarkers)
        } else {
            clearMarkersAnimated(existingMarkers)
        }
    }

    /// Clears the specified markers with an animation, moving them to the nearest new cluster if applicable.
    ///
    /// - Parameter markers: An array of `GMSMarker` objects to be removed with animation.
    ///
    func clearMarkersAnimated(_ markers: [GMSMarker]) {
        guard let mapView else { return }
        /// Remove existing markers: animate to nearest new cluster.
        let visibleBounds: GMSCoordinateBounds = GMSCoordinateBounds(region: mapView.projection.visibleRegion())

        for marker in markers {
            guard let userData = marker.userData else { return }
            /// If the marker for the attached userData has just been added, do not perform animation.
            guard !renderedClusterItems.contains(userData) else {
                marker.map = nil
                continue
            }

            /// If the marker is outside the visible view port, do not perform animation.
            guard !visibleBounds.contains(marker.position) else {
                marker.map = nil
                continue
            }

            guard let itemToNewClusterMap else { return }
            /// Find a candidate cluster to animate to.
            var toCluster: GMUCluster?
            if let cluster = marker.userData as? GMUCluster {
                toCluster = overlappingCluster(for: cluster, itemMap: itemToNewClusterMap)
            } else {
                let key = GMUWrappingDictionaryKey(object: userData)
                toCluster = itemToNewClusterMap[key]
            }
            
            /// If there is not near by cluster to animate to, do not perform animation.
            guard let toCluster else {
                marker.map = nil
                continue
            }

            /// All is good, perform the animation.
            CATransaction.begin()
            CATransaction.setAnimationDuration(animationDuration)
            marker.layer.latitude = toCluster.position.latitude
            marker.layer.longitude = toCluster.position.longitude
            CATransaction.commit()
        }
        
        // Clear existing markers after the animation has ended.
        DispatchQueue.main.asyncAfter(deadline: .now() + animationDuration * Double(NSEC_PER_SEC)) {
            self.clearMarkers(markers)
        }
    }

    /// Called when the camera is changed to reevaluate if new clusters need to be displayed
    /// because they have become visible.
    ///
    func update() {
        if let clusters {
            // Adds or updates clusters without animation.
            addOrUpdateClusters(clusters, animated: false)
        }
    }

    // MARK: - Private Methods
    /// Builds a lookup map that associates cluster items with their old or new clusters
    /// depending on whether the map is zooming in or zooming out.
    ///
    /// - Parameters:
    ///   - newClusters: An array of new `GMUCluster` objects.
    ///   - isZoomingIn: A boolean that indicates whether the map is zooming in (`true`) or out (`false`).
    ///
    func prepareClustersForAnimation(_ newClusters: [GMUCluster], isZoomingIn: Bool) {
        guard let mapView else { return }
        let zoom: Float = mapView.camera.zoom

        if isZoomingIn {
            itemToOldClusterMap = [:]
            guard let clusters else { return }
            for cluster in clusters {
                if !shouldRenderAsCluster(cluster, atZoom: zoom) && !shouldRenderAsCluster(cluster, atZoom: previousZoom) {
                    continue
                }
                for clusterItem in cluster.items {
                    let key = GMUWrappingDictionaryKey(object: clusterItem)
                    itemToOldClusterMap?[key] = cluster
                }
            }
            itemToNewClusterMap = nil
        } else {
            itemToOldClusterMap = nil
            itemToNewClusterMap = [:]
            for cluster in newClusters {
                if !shouldRenderAsCluster(cluster, atZoom: zoom) {
                    continue
                }
                for clusterItem in cluster.items {
                    let key: GMUWrappingDictionaryKey = GMUWrappingDictionaryKey(object: clusterItem)
                    itemToNewClusterMap?[key] = cluster
                }
            }
        }
    }

    /// Iterates through the given clusters and adds a marker for each cluster if:
    /// - The cluster is inside the visible region of the camera.
    /// - The cluster has not yet been added.
    ///
    /// - Parameters:
    ///   - clusters: An array of `GMUCluster` objects to be added or updated.
    ///   - animated: A boolean indicating whether the addition should be animated.
    ///
    func addOrUpdateClusters(_ clusters: [GMUCluster], animated: Bool) {
        guard let mapView else { return }
        let visibleBounds = GMSCoordinateBounds(region: mapView.projection.visibleRegion())

        for cluster in clusters {
            if renderedClusters.contains(cluster) {
                continue
            }

            var shouldShowCluster: Bool = visibleBounds.contains(cluster.position)
            let shouldRenderAsCluster: Bool = shouldRenderAsCluster(cluster, atZoom: mapView.camera.zoom)

            if !shouldShowCluster {
                for item in cluster.items {
                    if !shouldRenderAsCluster && visibleBounds.contains(item.position) {
                        shouldShowCluster = true
                        break
                    }
                    if animated {
                        let key: GMUWrappingDictionaryKey = GMUWrappingDictionaryKey(object: item)
                        if let oldCluster = itemToOldClusterMap?[key], visibleBounds.contains(oldCluster.position) {
                            shouldShowCluster = true
                            break
                        }
                    }
                }
            }

            if shouldShowCluster {
                renderCluster(cluster, animated: animated)
            }
        }
    }

    /// Renders the given cluster on the map. If the cluster should be rendered as a cluster, a cluster marker is added.
    /// Otherwise, individual items from the cluster are rendered.
    ///
    /// - Parameters:
    ///   - cluster: The `GMUCluster` to render.
    ///   - animated: A boolean indicating whether the rendering should be animated.
    ///
    func renderCluster(_ cluster: GMUCluster, animated: Bool) {
        guard let mapView else { return }
        let zoom = mapView.camera.zoom
        if shouldRenderAsCluster(cluster, atZoom: zoom) {
            var fromPosition: CLLocationCoordinate2D = kCLLocationCoordinate2DInvalid
            var animated: Bool = animated
            if animated {
                if let fromCluster = overlappingCluster(for: cluster, itemMap: itemToOldClusterMap) {
                    animated = true
                    fromPosition = fromCluster.position
                } else {
                    animated = false
                }
            }

            guard let icon: UIImage = clusterIconGenerator.iconForSize(cluster.count) else {
                return
            }
            let marker: GMSMarker = markerWithPosition(cluster.position, from: fromPosition, userData: cluster, clusterIcon: icon, animated: animated)
            mutableMarkers.append(marker)
        } else {
            for item in cluster.items {
                var marker: GMSMarker
                if let itemMarker = item as? GMSMarker {
                    marker = itemMarker
                    marker.map = mapView
                } else {
                    var fromPosition: CLLocationCoordinate2D = kCLLocationCoordinate2DInvalid
                    var shouldAnimate = animated
                    if shouldAnimate {
                        let key = GMUWrappingDictionaryKey(object: item)
                        if let fromCluster = itemToOldClusterMap?[key] {
                            fromPosition = fromCluster.position
                        }
                        shouldAnimate = true
                    } else {
                        shouldAnimate = false
                    }

                    marker = markerWithPosition(item.position, from: fromPosition, userData: item, clusterIcon: nil, animated: shouldAnimate)
                    if let title = item.title {
                        marker.title = title
                    }
                    if let snippet = item.snippet {
                        marker.snippet = snippet
                    }
                }
                mutableMarkers.append(marker)
                renderedClusterItems.add(item)
            }
        }
        renderedClusters.add(cluster)
    }

    /// Returns a `GMSMarker` for the given object.
    /// If the delegate implements `renderer(_:markerForObject:)`, it will use that method to obtain the marker.
    /// If no marker is provided by the delegate, a new `GMSMarker` is created and returned.
    ///
    /// - Parameter object: The object for which a marker is needed.
    /// - Returns: A `GMSMarker` for the object, either provided by the delegate or newly created.
    ///
    func marker(for object: Any) -> GMSMarker {
        if let marker = delegate?.renderer(self, markerForObject: object) {
            return marker
        }
        return GMSMarker()
    }

    /// Returns a marker at the final position of `position` with the attached `userData`.
    /// If `animated` is true, animates the marker from the specified `from` position.
    ///
    /// - Parameters:
    ///   - position: The final position of the marker.
    ///   - from: The initial position from which the marker should animate. If not animated, this is ignored.
    ///   - userData: The data to associate with the marker.
    ///   - clusterIcon: The icon for the cluster, if available. If nil, a default marker icon is used.
    ///   - animated: A boolean indicating whether the marker should animate to its position.
    /// - Returns: The created and rendered `GMSMarker`.
    ///
    func markerWithPosition(_ position: CLLocationCoordinate2D, from: CLLocationCoordinate2D, userData: Any, clusterIcon: UIImage?, animated: Bool) -> GMSMarker {
        let marker = marker(for: userData)
        let initialPosition: CLLocationCoordinate2D = animated ? from : position
        marker.position = initialPosition
        marker.userData = userData

        if let icon = clusterIcon {
            marker.icon = icon
            marker.groundAnchor = CGPoint(x: 0.5, y: 0.5)
        }
        marker.zIndex = Int32(zIndex)

        delegate?.renderer(self, willRenderMarker: marker)

        marker.map = mapView

        if animated {
            CATransaction.begin()
            CATransaction.setAnimationDuration(animationDuration)
            marker.layer.latitude = position.latitude
            marker.layer.longitude = position.longitude
            CATransaction.commit()
        }

        delegate?.renderer(self, didRenderMarker: marker)

        return marker
    }

    /// Returns clusters which should be rendered and are inside the camera's visible region.
    /// Filters the clusters based on their visibility and whether they should be rendered as clusters
    /// at the current zoom level.
    ///
    /// - Parameter clusters: The array of clusters to check for visibility.
    /// - Returns: An array of clusters that are inside the visible region and should be rendered.
    ///
    func visibleClusters(from clusters: [GMUCluster]) -> [GMUCluster]? {
        guard let mapView else { return nil }
        var visibleClusters = [GMUCluster]()
        let zoom = mapView.camera.zoom
        let visibleBounds = GMSCoordinateBounds(region: mapView.projection.visibleRegion())

        for cluster in clusters {
            if !visibleBounds.contains(cluster.position) { continue }
            if !shouldRenderAsCluster(cluster, atZoom: zoom) { continue }
            visibleClusters.append(cluster)
        }
        return visibleClusters
    }

    /// Returns the first cluster in the `itemMap` that shares a common item with the input `cluster`.
    /// This is used to heuristically find a candidate cluster to animate to or from.
    ///
    /// - Parameters:
    ///   - cluster: The input `GMUCluster` for which to find an overlapping cluster.
    ///   - itemMap: A dictionary mapping `GMUWrappingDictionaryKey` to `GMUCluster`.
    ///
    /// - Returns: The first overlapping `GMUCluster`, or `nil` if no overlap is found.
    ///
    func overlappingCluster(for cluster: GMUCluster, itemMap: [GMUWrappingDictionaryKey: GMUCluster]?) -> GMUCluster? {
        var found: GMUCluster?
        for item in cluster.items {
            let key: GMUWrappingDictionaryKey = GMUWrappingDictionaryKey(object: item)
            if let itemMap, let candidate: GMUCluster = itemMap[key] {
                found = candidate
                break
            }
        }
        return found
    }

    /// Clears the specified markers by removing them from the map and resetting their `userData`.
    ///
    /// - Parameter markers: An array of `GMSMarker` objects to be removed.
    ///
    func clearMarkers(_ markers: [GMSMarker]) {
        for marker in markers {
            // If the marker's userData conforms to the GMUCluster protocol, reset its userData.
            if marker.userData is GMUCluster {
                marker.userData = nil
            }
            marker.userData = nil
            // Remove the marker from the map.
            marker.map = nil
        }
    }

    /// Removes all existing markers from the attached map.
    /// Clears the `mutableMarkers` array, and also removes all rendered clusters, cluster items,
    /// and clears the cluster item maps (`itemToNewClusterMap`, `itemToOldClusterMap`).
    /// Resets the clusters to nil.
    ///
    func clear() {
        clearMarkers(mutableMarkers)
        mutableMarkers.removeAll()
        renderedClusters.removeAllObjects()
        renderedClusterItems.removeAllObjects()
        itemToNewClusterMap?.removeAll()
        itemToOldClusterMap?.removeAll()
        clusters = nil
    }

}
